insta/src/runtime.rs

```diff
@@ -623,25 +764,46 @@ pub fn assert_snapshot(
     expr: &str,
 ) -> Result<(), Box<dyn Error>> {
     let ctx = SnapshotAssertionContext::prepare(
-        refval,
+        &snapshot_value,
         workspace,
         function_name,
         module_path,
         assertion_file,
         assertion_line,
     )?;
 
-    // apply filters if they are available
-    #[cfg(feature = "filters")]
-    let new_snapshot_value =
-        Settings::with(|settings| settings.filters().apply_to(new_snapshot_value));
+    ctx.cleanup_previous_pending_binary_snapshots()?;
+
+    let content = match snapshot_value {
+        SnapshotValue::FileText { content, .. } | SnapshotValue::InlineText { content, .. } => {
+            // apply filters if they are available
+            #[cfg(feature = "filters")]
+            let content = Settings::with(|settings| settings.filters().apply_to(content));
+
+            let kind = match ctx.snapshot_file {
+                Some(_) => TextSnapshotKind::File,
+                None => TextSnapshotKind::Inline,
+            };
 
-    let kind = match ctx.snapshot_file {
-        Some(_) => SnapshotKind::File,
-        None => SnapshotKind::Inline,
+            TextSnapshotContents::new(content.into(), kind).into()
+        }
+        SnapshotValue::Binary {
+            content, extension, ..
+        } => {
+            assert!(
+                extension != "new",
+                "'new' is not allowed as a file extension"
```

Comment:
```suggestion
                "'.new' is not allowed as a file extension"
```

Sorry, mea culpa, I didn't think enough, reverting my suggestion